xds/src/main/java/io/grpc/xds/internal/sds/ServerWrapperForXds.java

```diff
@@ -77,27 +132,113 @@ public Server start() throws IOException {
+                future.set(throwable);
+                return;
+              }
+              xdsServingStatusListener.onNotServing(throwable);
+            } else {
+              // after start
+              if (isAbsent) {
+                xdsServingStatusListener.onNotServing(throwable);
+                if (currentServingState == ServingState.STARTED) {
+                  // shutdown the server
+                  delegate.shutdown();  // let existing calls finish on delegate
+                  currentServingState = ServingState.NOT_SERVING;
+                  delegate = null;
+                }
+              }
+            }
           }
 
           @Override
-          public void onSuccess() {
-            settableFuture.set(null);
+          public void onListenerUpdate() {
+            if (currentServingState == ServingState.STARTING) {
+              // during start()
+              future.set(null);
+            } else if (currentServingState == ServingState.NOT_SERVING) {
+              // coming out of NOT_SERVING
+              currentServingState = ServingState.ENTER_SERVING;
+              startRetryTask = new StartRetryTask();
+              startRetryTask.createTask(0L);
+            }
           }
         };
     xdsClientWrapperForServerSds.addServerWatcher(serverWatcher);
-    return settableFuture;
+    return future;
+  }
+
+  private final class StartRetryTask implements Runnable {
+
+    ScheduledExecutorService scheduledExecutorService;
+    ScheduledFuture<?> future;
+
+    private void createTask(long delay) {
+      if (scheduledExecutorService == null) {
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+      }
+      future = scheduledExecutorService.schedule(this, delay, timeUnitForDelayForRetry);
+    }
+
+    private void rebuildAndRestartServer() {
+      delegate = delegateBuilder.build();
+      try {
+        delegate = delegate.start();
+        currentServingState = ServingState.STARTED;
+        xdsServingStatusListener.onServing();
+        cleanUpStartRetryTask();
+      } catch (IOException ioe) {
+        xdsServingStatusListener.onNotServing(ioe);
+        if (isRetriableErrorInDelegateStart(ioe)) {
+          createTask(delayForRetry);
+        } else {
+          // permanent failure
+          currentServingState = ServingState.SHUTDOWN;
+          cleanUpStartRetryTask();
+        }
+      }
+    }
+
+    @Override
+    public void run() {
+      if (currentServingState != ServingState.ENTER_SERVING) {
+        throw new AssertionError("Wrong state:" + currentServingState);
+      }
+      rebuildAndRestartServer();
+    }
+
+    private void cleanUpStartRetryTask() {
+      if (scheduledExecutorService != null) {
+        scheduledExecutorService.shutdown();
+        scheduledExecutorService = null;
+      }
+      startRetryTask = null;
+    }
+
+    public void shutdownNow() {
+      if (future != null) {
+        future.cancel(true);
```

Comment:
I think `(true)` does not help much because there is nothing interruptible.

Reply:
The runnable contains a call to `delegate.start()` and that's interruptible - see [this](https://github.com/grpc/grpc-java/pull/7919/files#diff-8a5589de6831d1809fe656450f41d39a3daf65af4bd627f3b0df8386f6beb2f1R292)